packages/flame/lib/src/components/component.dart

```diff
@@ -255,63 +265,80 @@ class Component {
-  Future<void> add(Component component) async {
-    component.prepare(/*parent=*/ this);
-    if (component.isPrepared) {
-      // [Component.onLoad] (if it is defined) should only run the first time
-      // the component is added to a parent.
-      if (!component.isLoaded) {
-        final onLoadFuture = component.onLoad();
-        if (onLoadFuture != null) {
-          await onLoadFuture;
-        }
-        component.isLoaded = true;
-      }
-      // Should run every time the component gets a new parent, including its
-      // first parent.
-      component.onMount();
-      if (component.hasChildren) {
-        await component._reAddChildren();
-      }
-    }
-    children.addChild(component);
-  }
-
-  /// Adds multiple children.
+  /// The cost of this flexibility is that the component won't be added right
+  /// away. Instead, it will be placed into a queue, and then added later, after
+  /// it has finished loading, but no sooner than on the next game tick.
+  ///
+  /// When multiple children are scheduled to be added to the same parent, we
+  /// start loading all of them as soon as possible. Nevertheless, the children
+  /// will end up being added to the parent in exactly the same order as they
+  /// were originally scheduled by the user, regardless of how fast or slow
+  /// each of them loads.
   ///
-  /// See [add] for details.
+  /// A component can be added to a parent which may not be mounted to the game
+  /// tree yet. In such case, the component will start loading immediately, but
+  /// its mounting will be delayed until such time when the parent becomes
+  /// mounted.
+  ///
+  /// This method returns a future that completes when the component is done
+  /// loading, and mounting if the parent is currently mounted. However, this
+  /// future will not guarantee that the component will become "fully mounted":
+  /// it still needs to be added to the parent's children list, and that
+  /// operation will only be done on the next game tick.
+  ///
+  /// A component can only be added to one parent at a time. It is an error to
+  /// try to add it to multiple parents, or even to the same parent multiple
+  /// times. If you need to change the parent of a component, use [changeParent]
+  /// method.
+  Future<void> add(Component component) => component.addToParent(this);
+
+  /// A convenience method to [add] multiple children at once.
   Future<void> addAll(Iterable<Component> components) {
     return Future.wait(components.map(add));
   }
 
-  /// The children are added again to the component set so that [prepare],
-  /// [onLoad] and [onMount] runs again. Used when a parent is changed
-  /// further up the tree.
-  Future<void> _reAddChildren() async {
-    if (_children != null) {
-      await Future.wait(_children!.map(add));
-      await Future.wait(_children!.addLater.map(add));
+  /// Adds this component to the provided [parent] (see [add] for details).
+  Future<void> addToParent(Component parent) async {
```

Comment:
Maybe this should be private? So that we try to get people to use one way of adding children, since both methods require both the child and the parent anyways?

Reply:
My thinking was that since we have both `remove` and `removeFromParent`, there should also be `add` and `addToParent` -- for symmetry.

Reply:
Hmm, I see your point, but the `removeFromParent` is actually making thinga simpler in some cases since you don't need to know the parent, this one just flips around the argument and the instance it called on. Let's see what the others say.

Reply:
Well, this function can be useful too. For example, you can write
```dart
SomeComponent()
  ..position = Vector2(x, y)
  ..scale = Vector2.all(1.5)
  ..addToParent(level);
```